overlord/snapstate/snapmgr.go

```diff
@@ -989,6 +990,79 @@ func (m *SnapManager) blockedTask(cand *state.Task, running []*state.Task) bool
 	return false
 }
 
+func affectsRunningHooks(cand *state.Task, running []*state.Task) (block bool) {
+	st := cand.State()
+
+	affectingTasks := map[string][]state.Status{
+		"unlink-current-snap": {state.DoStatus, state.DoingStatus},
+		// this will block any undo link-snap, although only undos of first installs/snap
+		// unlink. It's unlikely that a hook runs between the link-snap and the end
+		// of the change and, in the worst case, we delay the undoing
+		"link-snap":   {state.UndoStatus, state.UndoingStatus},
+		"unlink-snap": {state.DoStatus, state.DoingStatus},
+	}
+
+	var unlinkingTask bool
+out:
+	for kind, statuses := range affectingTasks {
+		if cand.Kind() != kind {
+			continue
+		}
+
+		for _, status := range statuses {
+			if cand.Status() == status {
+				unlinkingTask = true
+				break out
+			}
+		}
+	}
```

Comment:
would this work?

```suggestion
	statuses, ok := affectingTasks[cand.Kind()]
	if !ok {
		return false
	}
	
	for _, status := range statuses {
		if cand.Status() == status {
			unlinkingTask = true
			break
		}
	}
```

you could also do a trick like this:
```
	type unlinkingTask struct {
		kind string
		status state.Status
	}
	affectingTasks := map[unlikingTask]bool{
		{"unlink-current-snap", state.DoStatus}: true,
		{"unlink-current-snap", state.DoingStatus}: true,
		// this will block any undo link-snap, although only undos of first installs/snap
		// unlink. It's unlikely that a hook runs between the link-snap and the end
		// of the change and, in the worst case, we delay the undoing
		{"link-snap", state.UndoStatus}: true,
		{"link-snap", state.UndoingStatus}: true,
		{"unlink-snap", state.DoStatus}: true,
		{"unlink-snap", state.DoingStatus}: true,
	}

	if _, ok := affectingTasks[unlinkingTask{cand.Kind(), cand.Status()}]; !ok {
		return false
	}
```